lib/TPP/Transforms/ConvertPackToExpandShape.cpp

```diff
@@ -0,0 +1,162 @@
+    // TODO: is it the case that each dim can only occur once in innerDimPos?
+    if (llvm::is_contained(innerDimPos, idx))
+      associationIndices.back().push_back(curDimIdx++);
+  }
+
+  rewriter.setInsertionPointAfter(packOp);
+  auto expandShape = rewriter.create<tensor::ExpandShapeOp>(
+      packOp->getLoc(), normalizedType, packOp.getOperand(0),
+      ArrayRef(associationIndices));
+
+  return std::pair(expandShape, normalizedIndexingMap);
+}
+
+static FailureOr<tensor::CollapseShapeOp>
+unpackToCollapseShape(tensor::UnPackOp unpackOp, PatternRewriter &rewriter) {
+  auto origType =
+      dyn_cast<TensorType>(unpackOp->getResult(0).getType());
+  auto origShape = origType.getShape();
+  auto innerDimPos = SmallVector<unsigned int>(unpackOp.getInnerDimsPos());
+
+  SmallVector<SmallVector<int64_t, 2>> associationIndices;
+  int curDimIdx = 0;
+  for (auto idx : llvm::seq(origShape.size())) {
+    associationIndices.emplace_back(SmallVector<int64_t>());
+    associationIndices.back().push_back(curDimIdx++);
+    // TODO: is it the case that each dim can only occur once in innerDimPos?
+    if (llvm::is_contained(innerDimPos, idx))
+      associationIndices.back().push_back(curDimIdx++);
+  }
+
+  rewriter.setInsertionPointAfter(unpackOp);
+  auto collapseShape = rewriter.create<tensor::CollapseShapeOp>(
+      unpackOp.getLoc(), origType, unpackOp.getOperand(0),
+      ArrayRef(associationIndices));
+
+  return collapseShape;
+}
+
+struct ConvertPackToExpandShape : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!linalg::isaContractionOpInterface(genericOp))
+      return failure();
+    // know linalg has two inputs and one output and is a contraction
+
+    auto indexingMaps = genericOp.getIndexingMapsArray();
+    // TODO: need way to control which operands to revert packing on
+    auto idxSet = {0, 1, 2};
+    Type resultType = nullptr;
+    for (auto idx : idxSet) {
+      auto packOp = dyn_cast_if_present<tensor::PackOp>(
+          genericOp->getOperand(idx).getDefiningOp());
+      if (!packOp)
+        return failure();
+
+      auto res = packToExpandShape(packOp, indexingMaps[idx], rewriter);
+      if (!succeeded(res))
+        return res;
+
+      rewriter.replaceAllOpUsesWith(packOp, res->first);
+      if (idx == 2) {
+        resultType = res->first.getResultType();
+        genericOp->getOpResult(0).setType(resultType);
+      }
+
+      SmallVector<Attribute> indexingMaps =
+          llvm::to_vector(genericOp.getIndexingMaps());
+      indexingMaps[idx] = AffineMapAttr::get(res->second);
+
+      genericOp.setIndexingMapsAttr(
+          ArrayAttr::get(rewriter.getContext(), indexingMaps));
+    }
+
+    if (auto unpackOp = llvm::dyn_cast<tensor::UnPackOp>(
```

Comment:
I'd move it to a separate pattern

Reply:
On the one hand, yes, it might be nice to have it available as a separate pattern. On the other hand, when we modify the out parameter of the linalg, the result type changes/needs to be changed as well. This in turn means the unpack afterwards needs to have _something_ done to it. In the context of a single pattern it is clear what this _something_ is: rewrite it to a collapse_shape. 

In case of separate patterns, do you have an idea what to do with the result of the linalg when we change the shape of the out parameter? If we don't do anything the IR is most likely left in an illegal state (i.e. types will not match up).

Reply:
When it is not the case that there is a single unpack user of the linalg, we probably shouldn't do the rewrite of the out parameter any way (as we don't know how the users are impacted by the shape change of the type). The only sensible thing seems to be to insert some op in between the linalg's result and all the users to invert the shape change.

Reply:
For outs, we definitely need the whole chain pack->generic->unpack. If there's any user other than unpack then it probably doesn't make sense to change outs shape.
Also this case might require extra checks on the pack-unpack pair to see if they are compatible.

So my thinking is mainly that it'd be cleaner to move this extra logic to a separate pattern that focuses on outs, and this one focuses on ins. But maybe it doesn't make much difference.

Reply:
Done.

Reply:
If you could give an example of a compatibility issue for pack-unpack pairs, that would much appreciated. I had a think and came to the conclusion that as the generic's result type is shared by the pair, this is enough to force compatability.

Reply:
Not sure if you can fold a pair in which pack has different perms than unpack `outer_dims_perm = [1, 0] inner_dims_pos = [0, 1]`.

Reply:
There are also probably cases when you can fold pack but need to keep unpack. For example, maybe pack can have permutation which then is performed by the generic but unpack has to stay as you need extra data movement anyway.

But let's ignore these for now.

Reply:
Moved to two separate patterns. There's discussion below as to the safe permutations on unpack.